LineageOS 18.1

hmenke opened the pull request:
https://lineageos.org/Changelog-25/

danielfullmer commented:
Thanks! Looks fairly straightforward.

There's one thing I was hoping we could do (in a future PR if you'd rather not deal with it now):
Have both lineageos-17.1 and lineageos-18.1 available, and have robotnix automatically choose the default `androidVersion` for a particular device based on the `lineage-build-targets` from the `hudson` repo.  That way, we keep supporting any old devices until they're updated, and also allow the user the option to manually set `androidVersion=11` if they want to try out the experimental branch for their device.

lapp0 commented:
Thanks for your great work hmenke! I'm running 18.1 with this code. 18.1 is still a bit buggy and missing things like xposed that are only available for android 10 right now. Perhaps there should be a `lineageos-18.1` branch in this repo as well so users can choose between 17.1 and 18.1?

hmenke commented:
@danielfullmer Done. In the previous LineageOS update you mentioned

> Tested [...] by running checkAndroid on all 150 lineageos devices.

Do you have a script for that, so that I can do that too for 18.1?

danielfullmer commented on flavors/lineageos/default.nix:
```diff
@@ -49,11 +49,11 @@ let
 in mkIf (config.flavor == "lineageos")
 {
-  androidVersion = mkDefault 10;
+  androidVersion = mkDefault 11;
```
I actually mean that we would set the default `androidVersion` here based on what device the user had chosen, and using information from `lineage-build-targets` which should end up in `device-metadata.json`.

danielfullmer commented on flavors/lineageos/update.sh:
```diff
@@ -21,7 +21,7 @@ fi
     --ref-type branch
     "https://github.com/LineageOS/android"
-    "lineage-17.1"
+    "lineage-18.1"
```
Since we would want to allow the user to update each branch independently, we should take the branch as an argument to the script.

danielfullmer reviewed (changes requested):
The updater scripts would need some changes to support both `lineage-17.1` and `lineage-18.1`. A couple things obvious things I can see: `update-device-metadata.py` should produce `device-metadata.json` with devices supporting either branch, instead of filtering for just one branch.  The `update-device-dirs.py` script should take the branch as an argument, and `device-dirs.json` and `vendor-dirs.json` should be split into files supporting the 17.1 and 18.1 branches, with the nix expression choosing which one to use.

danielfullmer commented:
@hmenke Here's what I had used in the past:

```nix
let
  pkgs = import ./pkgs {};
  lib = pkgs.lib;
  deviceMetadata = lib.importJSON ./flavors/lineageos/device-metadata.json;
in
  map (device: (import ./default.nix {
    inherit pkgs;
    configuration = {
      inherit device;
      flavor="lineageos";
    };
  }).config.build.checkAndroid ) 

(lib.attrNames (lib.filterAttrs (n: v: v.branch == "lineage-17.1") deviceMetadata))
```

Preferably we'd remove the filterAttrs part so we'd check all devices using their supported branch.

Kranzes commented:
How is this PR going so far?, is anyone still working on it?

danielfullmer commented:
I've just pushed up a new `lineageos-18.1` branch with a few extra commits on top of @hmenke 's branch: https://github.com/danielfullmer/robotnix/commits/lineageos-18.1

The notable one is this: https://github.com/danielfullmer/robotnix/commit/f859b08a4d3d41076b1f85e51ec546b3292a8168
The idea is that we only need one `device-metadata.json` file, not one for each branch.  This metadata file contains the `branch` attribute for each device, which says which branch we should use by default for a build. The other big change in that commit is that the updater scripts are able to fetch both branches for each device.  This should allow the user to set `androidVersion=11` to try out `lineage-18.1` if that branch is available, even if it hasn't yet been promoted to the default branch for that device.

I'm currently building `config.build.checkAndroid` output for each of the 166 devices.  I'm also going to try to build and test both `lineage-18.1` and `lineage-17.1` for `sunfish.

@Kranzes  Thanks for the reminder. I had intended to look more closely at this, but got distracted with other things.
@hmenke Thank you very much for your work on this, and I'm open to any comments/questions about the follow up commits. We might also want to rebase/squash some commits before merging into master.

danielfullmer commented:
Update: `sunfish` worked on `lineage-18.1`.  Although there is a branch for `lineage-17.1` on sunfish, it doesn't look like it was ever supported on 17.1, and so it doesn't build at all.  So, unfortunately, I don't have any free devices that support both 17.1 and 18.1.  (`crosshatch` does, but it's my daily driver).

danielfullmer commented:
All 166 devices were able to build `config.build.checkAndroid` output successfully. So, at least we know we're not missing any obvious dependencies for these devices.

hmenke commented:
I've fixed the evaluation error, followed your recommendation of unifying `device-metadata.json` and rebased onto `master`. It looks like on your `lineageos-18.1` branch you have already fixed all the updater scripts. Please merge them into this PR and just push to my branch.

danielfullmer commented:
@hmenke Rebased and pushed changes to your branch.

Unless there are any issues I've missed. I think this should be good to merge. If anyone has another device to that they'd like to test LineageOS 18.1 with, it would be nice to confirm that it works for others.
